File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        
        
        int[] leftb=new int[heights.length];
        
        Stack<Integer> st=new Stack<>();
        
        // st.push(heights.length - 1);
        
        // leftb[heights.length-1]=heights.length;
        
        for(int i=heights.length-1;i>=0;i--){
            
            while(st.size()>0 && heights[i]<=heights[st.peek()]){
                st.pop();
            }
            if(st.size()==0){
                leftb[i]=heights.length;
            }
            else{
                leftb[i]=st.peek();
            }
            
            st.push(i);
        }
        
      // _______________________________________________________
         int[] rightb=new int[heights.length];
        
        Stack<Integer> st2=new Stack<>();
        
        // rightb[0]=-1;
        
        for(int i=0;i<heights.length;i++){
            
            while(st2.size()>0 && heights[i]<=heights[st2.peek()]){
                st2.pop();
            }
            if(st2.size()==0){
                rightb[i]=(-1);
            }
            else{
                rightb[i]=(st2.peek());
            }
            
            st2.push(i);
        }
        
        // ____________________________________________________________
    
        
        int max=0;
        for(int i=0;i<heights.length;i++){
            
            int width=(leftb[i]-rightb[i])-1;
            int area=heights[i] * width;
            max=Math.max(area,max);
            
            // System.out.print(rightb[i]);
        }
        
        // System.out.print(max);
        return max;
    }   
}

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.js
var largestRectangleArea = function(arr) {
    arr.push(0)
    const stack = [];
    let maxArea = 0;
    let curr;
    let currH;
    let top; 
    let topH; 
    let area;

    for(let i = 0; i < arr.length; i++) {
        top = stack[stack.length-1];
        topH = arr[top];
        while(stack.length > 1 && topH > arr[i]) {
            curr = stack.pop();
            currH = arr[curr];
            top = stack[stack.length-1];
            topH = arr[top];
            area = currH * (i - 1 - top);
            maxArea = Math.max(area, maxArea);
        }
        
        if(stack.length && topH > arr[i]) {
            curr = stack.pop();
            currH = arr[curr];
            area = currH * i;
            maxArea = Math.max(area, maxArea);
        }
        stack.push(i);
    }
    return maxArea;
};